### PR TITLE
#10783 - Fix for Get-AzResource returning PSResource with null SubscriptionId

### PR DESCRIPTION
**Pull request by JohnLBevan**
## Description
To resolve issue: https://github.com/Azure/azure-powershell/issues/10783
Ensure SubscriptionId is not left null when initialising the PSResource object / using Get-AzResource cmdlet.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by msJinLei** [comment]
@JohnLBevan are you msft? Please join  https://github.com/Azure

File: src/Resources/Resources/ChangeLog.md [hunk 1]
```diff
@@ -38,7 +39,7 @@
 
 ## Version 3.4.1
 * Added upcoming breaking change warnings on below cmdlets, because the value of `IdentifierUris` parameter will need verified domain.
-  - `New-AzADApplication` 
+  - `New-AzADApplication`
```
Comment (msJinLei) [inline src/Resources/Resources/ChangeLog.md:42]: Please don't change the previous log

File: src/Resources/Resources/ChangeLog.md [hunk 2]
```diff
@@ -73,7 +74,7 @@
 
 ## Version 3.1.0
 * Added `-Tag` parameter support to `Set-AzTemplateSpec` and `New-AzTemplateSpec`
-* Added Tag display support to default formatter for Template Specs 
+* Added Tag display support to default formatter for Template Specs
```
Comment (msJinLei) [inline src/Resources/Resources/ChangeLog.md:77]: Please don't change the previous log

File: src/Resources/Resources/ChangeLog.md [hunk 3]
```diff
@@ -91,7 +92,7 @@
 * Fixed an issue where template deployment cmdlets does not preserve case for template parameters
 * Added a default API version to be used in `Export-AzResourceGroup` cmdlet
 * Added cmdlets for Template Specs (`Get-AzTemplateSpec`, `Set-AzTemplateSpec`, `New-AzTemplateSpec`, `Remove-AzTemplateSpec`, `Export-AzTemplateSpec`)
-* Added support for deploying Template Specs using existing deployment cmdlets (via the new -TemplateSpecId parameter) 
+* Added support for deploying Template Specs using existing deployment cmdlets (via the new -TemplateSpecId parameter)
```
Comment (msJinLei) [inline src/Resources/Resources/ChangeLog.md:95]: Please don't change the previous log

File: src/Resources/Resources/ChangeLog.md [hunk 4]
```diff
@@ -177,7 +178,7 @@
     - Remove-AzTag -ResourceId
 * Added new Tag cmdlet
     - Update-AzTag -ResourceId
-* Brought ScopedDeployment from SDK 3.3.0 
+* Brought ScopedDeployment from SDK 3.3.0
```
Comment (msJinLei) [inline src/Resources/Resources/ChangeLog.md:181]: Please don't change the previous log

**Review by msJinLei — CHANGES_REQUESTED** [review]
Can you add a check of subscription id after your change applied?

**Comment by JohnLBevan** [comment]
> @JohnLBevan could you join https://github.com/Azure?

I'm happy to, but not sure how without an invite.  Please could you add me as a member to trigger that invite, or else point me to info on how to join?  Thanks in advance.

> Can you add a check of subscription id after your change applied?

Sure, will look into that shortly...

> Please don't change the previous log

Apologies; my VSCode settings automatically trim whitespace on markdown line endings.  Will review the commits to only submit deliberate changes in future.

File: src/Resources/Resources/ChangeLog.md (hunk 2 above)
Comment (JohnLBevan) [reply src/Resources/Resources/ChangeLog.md]: Reverted in c858365

File: src/Resources/Resources/ChangeLog.md (hunk 3 above)
Comment (JohnLBevan) [reply src/Resources/Resources/ChangeLog.md]: Reverted in c858365

File: src/Resources/Resources/ChangeLog.md (hunk 1 above)
Comment (JohnLBevan) [reply src/Resources/Resources/ChangeLog.md]: Reverted in c858365

File: src/Resources/Resources/ChangeLog.md (hunk 4 above)
Comment (JohnLBevan) [reply src/Resources/Resources/ChangeLog.md:181]: Reverted in c858365

**Comment by JohnLBevan** [comment]
> Can you add a check of subscription id after your change applied?

Resolve in commit 1afea3b.

**Comment by msJinLei** [comment]
> @msJinLei
> 
> > @JohnLBevan could you join https://github.com/Azure?
> 
> I'm happy to, but not sure how without an invite. Please could you add me as a member to trigger that invite, or else point me to info on how to join? Thanks in advance.
> 
Sorry, are you a msft? if so, please link your github account with your microsoft account https://docs.opensource.microsoft.com/tools/github/index.html

**Comment by JohnLBevan** [comment]
> Sorry, are you a msft?

Ah - sadly not.

**Comment by msJinLei** [comment]
> > Sorry, are you a msft?
> 
> Ah - sadly not.

Welcome to contribute! I will involve domain review here.

**Comment by msJinLei** [comment]
@ifeoluwaokunoren could you help to review the PR? thanks

**Comment by BethanyZhou** [comment]
Hi @Grayer123 , @MiYanni and @cormacpayne , is there anyone can help to review this PR?

File: src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs [hunk 5]
```diff
@@ -77,96 +77,92 @@ public string TagsTable
...
-            this.Name = resource.Name;
-            this.Plan = resource.Plan;
-            this.Properties = ((JToken)resource.Properties).ToPsObject();
-            this.ResourceType = resource.Type;
-            this.Sku = resource.Sku;
-            this.Tags = resource.Tags;
-            this.Type = resource.Type;
-
-            var resourceIdentifier = new ResourceIdentifier(this.Id);
-            this.ParentResource = resourceIdentifier.ParentResource;
```
Comment (isra-fel) [inline src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs:98]: ParentResource is missing.

File: src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs [hunk 6]
```diff
@@ -77,96 +77,92 @@ public string TagsTable
...
+            this.Plan = plan;
+            this.Sku = sku;
+            this.Tags = tags;
         }
 
-        public PSResource(GenericResourceExpanded resource)
+        public PSResource(GenericResource resource): this(
+            resource.Id,
+            (new ResourceIdentifier(resource.Id)).Subscription, 
+            (new ResourceIdentifier(resource.Id)).ResourceGroupName,
```
Comment (isra-fel) [inline src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs:102]: If we add `ParentResource` here would be 3 `new`s. I'm thinking we should move in down in the constructor body.

**Review by isra-fel — CHANGES_REQUESTED** [review]
Thanks for contributing! What you did generally looks good to me.
Please see the inline comments 😀

File: src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs (hunk 5 above)
Comment (JohnLBevan) [reply src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs:98]: Good spot / apologies.  Corrected ( https://github.com/JohnLBevan/azure-powershell/blob/issue-10783-PSResource/src/Resources/ResourceManager/SdkModels/Resources/PSResource.cs#L113 )

**Review by isra-fel — APPROVED** [review]
Thanks again for your contribution!
I'll merge the pull request and it will be in our next official release 😁

**Pull request merged**